hail/python/hail/matrixtable.py

```diff
@@ -2807,25 +2807,205 @@ def write(
+                block_right_align = right_align[block_slice]
+                hline = format_hline(block_display_column_widths)
+
+                ascii_str += hline
+                ascii_str += format_line(
+                    display_column_names[block_slice], block_display_column_widths, block_right_align
+                )
+                ascii_str += hline
+                if should_show_types:
+                    ascii_str += format_line(
+                        truncated_types[block_slice], block_display_column_widths, block_right_align
+                    )
+                    ascii_str += hline
+                for row in truncated_values:
+                    ascii_str += format_line(row[block_slice], block_display_column_widths, block_right_align)
+                ascii_str += hline
+
+            if n_rows < matrix_table.count_rows():
+                row_count = len(rows)
+                ascii_str += f"showing top { row_count } { 'row' if row_count == 1 else 'rows' }\n"
+            total_n_cols = matrix_table.count_cols()
+            if n_cols != total_n_cols:
+                ascii_str += f"showing the first { n_cols } of { total_n_cols } columns"
+
+            return ascii_str
 
         def __repr__(self):
             return self.__str__()
 
         def _repr_html_(self):
-            s = self.table_show._repr_html_()
-            if self.displayed_n_cols != self.actual_n_cols:
+            # This method is not thoroughly tested and should be used carefully.
+            total_n_cols = self.matrix_table.count_cols()
+            table = self.matrix_table.localize_entries('entries', 'cols')
+            if len(table.key) > 0:
+                table = table.order_by(*table.key)
+            col_key_type = self.matrix_table.col_key.dtype
+
+            col_headers = [f'<col {i}>' for i in range(0, self.n_cols)]
+            if len(col_key_type) == 1 and col_key_type[0] in (hl.tstr, hl.tint32, hl.tint64):
+                cols = self.matrix_table.col_key[0].take(self.n_cols)
+                if len(set(cols)) == len(cols):
+                    col_headers = [repr(c) for c in cols]
+
+            entries = {col_headers[i]: table.entries[i] for i in range(0, self.n_cols)}
+            table = table.select(
+                **{f: table[f] for f in self.matrix_table.row_key},
+                **{f: table[f] for f in self.matrix_table.row_value if self.include_row_fields},
+                **entries,
+            )
+            table_show = table._show(self.n_rows, self.width, self.truncate_limit, self.should_show_types)
+            s = table_show._repr_html_()
+            if self.n_cols != total_n_cols:
                 s += '<p style="background: #fdd; padding: 0.4em;">'
-                s += f"showing the first { self.displayed_n_cols } of { self.actual_n_cols } columns"
+                s += f"showing the first { self.n_cols } of { total_n_cols } columns"
                 s += '</p>\n'
             return s
```

Comment:
Not necessary for this PR (which hugely improves the ASCII table!), but I think it would be quite simple to extend your approach for an ASCII table to producing an HTML table. Easier, even, because you don't need to worry about cooking up the padding of the ASCII cells. You just need to put the HTML tags in the right places. We could probably even [delegate truncation to HTML/CSS](https://stackoverflow.com/questions/5239758/how-can-i-truncate-table-cells-but-fit-as-much-as-content-possible).